aiida/backends/sqlalchemy/models/group.py

```diff
@@ -47,6 +47,8 @@ class DbGroup(Base):
     time = Column(DateTime(timezone=True), default=timezone.now)
     description = Column(Text, nullable=True)
 
+    extras = Column(JSONB, default=dict, server_default='{}')
```

Comment:
Why do we need the `server_default='{}'` here, but not for the `attributes` nor `extras` of `DbNode`?

Reply:
For the new `extras` of the `Group`, we need to add this column with an empty dictionary to the database in the migration. If I don't add `server_default` in the column declaration, the SqlAlchemy tests fail because the model and database schemas do not match (i.e. the `test_model_and_migration_schemas_are_the_same` test fails; see [here](https://github.com/mbercx/aiida-core/runs/1111568831?check_suite_focus=true)). 

Adding the `server_default` setting for the `DbNode` columns causes similar issues in the testing, and more importantly lead to issues migrating the database from the previous schema. The first can be fixed by manually adding `alter_column` statements to the SqlAlchemy migration file (see [here](https://github.com/mbercx/aiida-core/actions/runs/243324072)), but when migrating from the previous schema the migrate process got stuck. I'll see if I can reproduce this issue.

Reply:
Hmm, I no longer seem to be able to reproduce the issue I was having with the SqlAlchemy database migration. So, I'd vote for adding `server_default='{}'` to the `attributes` and `extras` of the `DbNode` model as well, and adding the `alter_column` statements to the SqlAlchemy migration file. Note that this *should* not have any influence for the users. Any node created with either the previous or current schema will set the `attributes` and `extras` to an empty dictionary upon initialization, so it will always be stored with this default in the database. I'd mainly change it for consistency.

Note again that removing the `server_default` everywhere is not an option, since this would mean groups from previous schemas would have a `null` value for the `extras` in the database upon migration. The current methods for the `extras` expect a dictionary in the database, however, so this leads to errors.

Reply:
Adding the `server_default` to the `DbNode` model will require a schema migration though. The `server_default` is included in the schema and so the database will be responsible for generating it. Contrastingly, the `default` is generated by the ORM in Python. By default SqlAlchemy does not populate the defaults of a model upon instantiation, but only when it is persisted. The reason for this slightly unintuitive behavior is that, depending on the nature of the default, it cannot always be _known_ before it is persisted. For consistency, it also does this for normal `default` values, which _could_ be known. There is a listener `from sqlalchemy_utils import force_instant_defaults` that can be called to force these defaults being set, which is what we actually do in `aiida.backends.sqlalchemy.model.__init__.py`. Finally, setting _both_ `default` and `server_default` is not necessary and can only cause problems.

I think then, that it is best to just keep the `default` and forget about the `server_default`. Anyway, generating an empty dictionary in Python and sending it to Postgres, instead of having PostgreSQL take care of it, should not matter much in terms of performance. Together with the force defaults listener, this allows us to have the defaults automatically set upon model instantiation. The only remaining problem you faced was the failing test, but the correct solution there is just to update the automatically generated schema and remove the `server_default` from the `add_column` statement. I think this should answer all your questions, but I may have missed something, in which case please let me know.

I also think that this would be the last open question for this PR correct? Except for the creation of a `ExtrasMixin` for the front-end ORM. I would suggest you make the new PR that refactors the backend classes with the new mixins and you also add the front-end ORM mixin. Then we can merge that, and we can rebase this PR. Sound good?

Reply:
Thanks for the explanation! I agree that setting both the `default` and `server_default` in the `Column` declaration is not necessary (I hadn't noticed the `force_instant_defaults` function though, good to know about this subtlety). However, if I do not add `server_default='{}'` to the `add_column` statement, already existing groups will get a `null` value in the `extras` column upon migration. This means `Group.extras` will return None, which creates issues when you try to e.g. use the `BackendEntityExtrasMixin.set_extra()` method, since this expects the `extras` to be a dictionary. 

One solution would be to accommodate for `null` values in the database, but doing this just because of the migration issue is probably not worth it. So perhaps it's best to just leave the `server_default` for the group extras, both in the `add_column` statement and column declaration?

> I would suggest you make the new PR that refactors the backend classes with the new mixins and you also add the front-end ORM mixin. Then we can merge that, and we can rebase this PR. Sound good?

Yes! I'll start working on a new PR that adds the mixin classes, also for the front-ends.

Reply:
> However, if I do not add server_default='{}' to the add_column statement, already existing groups will get a null value in the extras column upon migration. This means Group.extras will return None, which creates issues when you try to e.g. use the BackendEntityExtrasMixin.set_extra() method, since this expects the extras to be a dictionary.

The real fix is to have a different, slightly more complicated strategy:
1. Add the column without the `server_default`
2. Set an empty dictionary for all existing groups (either in through ORM or directly in SQL)

Note that this will work because you also mark the new column as `nullable=True`. Also this does not really make sense if you have a default. So I think we should make it non nullable. This will prevent us accidentally inserting rows with null values, despite setting a default in the Python ORM. Unfortunately, this does make the migration yet again a bit more complex

1. Add the column without the `server_default` and `nullable=True`
2. Set an empty dictionary for all existing groups (either in through ORM or directly in SQL)
3. Update the column to `nullable=False`

You can see the migration `041a79fc615f_dblog_cleaning.py` for an example where a similar problem had to be addressed.